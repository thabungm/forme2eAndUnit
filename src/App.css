.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

.container {
  background: gray;
  padding: 10px;
}
.inputEl {
  padding: 5px;
  width: 100%;
}
form {
  width:500px;
}

body{
  display:flex;
  justify-content: center;
}
.label, .error {
  text-align: left;
  padding: 5px;
}
.error {
  font-size: 13px;
  padding-bottom: 6px;
  color: #d32f2f;
}
button {
  padding: 10px;
  width: 100px;
}

.success {
  color: #0080008c;
  padding: 10px;
  margin: 10px;
  border-radius: 2px;
  /* color: white; */
}