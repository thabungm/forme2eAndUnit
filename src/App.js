import './App.css';
import { StyledEngineProvider } from '@mui/material/styles';
import Form from './components/Form';
import { QueryClient, QueryClientProvider } from 'react-query';

function App() {
  const queryClient = new QueryClient();

  return (
    <StyledEngineProvider>
      <QueryClientProvider client={queryClient}>
        <Form />
      </QueryClientProvider>
    </StyledEngineProvider>
  );
}

export default App;
