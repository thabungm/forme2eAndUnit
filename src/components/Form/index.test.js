import { shallow, mount } from 'enzyme';

import App from './index';
let mockData = {
  mutation: {
    isLoading: false,
    error: false,
    data: [
      {
        id: 1,
      },
      {
        id: 2,
      },
    ],
    mutateAsync: jest.fn(),
  },
  query: {
    isLoading: false,
    error: false,
    data: [
      {
        id: 1,
      },
      {
        id: 2,
      },
    ],
  },
};

function resetData() {
  // mockData = JSON.parse(JSON.stringify(defaultData));
}

jest.mock('react-query', () => {
  return {
    useQuery: () => mockData.query,
    useMutation: () => mockData.mutation,
  };
});

describe('Renders snapshot', () => {
  it('SHOULD render the form with components', () => {
    const result = shallow(<App />);
    expect(result).toMatchSnapshot();
    expect(result.find('h3').length).toBe(1);
    expect(result.find('FormTextInput').length).toBe(2);
    expect(result.find('FormControl').length).toBe(1);
  });
});

describe('When API is loading', () => {
  afterEach(() => {
    resetData();
  });
  it('SHOULD show loading', () => {
    mockData.query.isLoading = true;
    const result = shallow(<App />);
    expect(result).toMatchSnapshot();
    expect(result.find('div').text()).toBe('Loading');
  });
});

describe('When form is submitted', () => {
  beforeEach(() => {
    mockData.query.isLoading = false;
  });
  describe('AND when the fields are NOT filled', () => {
    it('SHOULD show error', async () => {
      const result = shallow(<App />);
      expect(result).toMatchSnapshot();

      const form = result.find('form');
      await form.simulate('submit');
      console.log(mockData.mutation);
      expect(mockData.mutation.mutateAsync).toHaveBeenCalledTimes(0);
    });
  });

  describe('AND when the fields are filled', () => {
    it('SHOULD show error', async () => {
      mockData.query.isLoading = false;

      const result = shallow(<App />);
      expect(result).toMatchSnapshot();

      const form = result.find('form');
      await form.simulate('submit');
      console.log(mockData.mutation);
      expect(mockData.mutation.mutateAsync).toHaveBeenCalledTimes(0);
    });
  });
});
