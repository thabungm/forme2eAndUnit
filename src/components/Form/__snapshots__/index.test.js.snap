// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders snapshot SHOULD render the form with components 1`] = `
<Fragment>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <h3>
      Form Demo
    </h3>
    <div
      className="success"
    />
    <FormControl
      errorMessage="Please select one"
      label="Choose userId"
      name="userId"
      onChange={[Function]}
      required={true}
      value=""
    />
    <FormTextInput
      fullWidth={true}
      label="Enter title"
      name="title"
      onChange={[Function]}
      required={true}
      value=""
      variant="outlined"
    />
    <FormTextInput
      fullWidth={true}
      label="Enter body"
      name="body"
      onChange={[Function]}
      required={true}
      value=""
    />
    <ForwardRef(Button)
      disabled={false}
      type="submit"
      variant="outlined"
    >
      Submit
    </ForwardRef(Button)>
  </form>
  <Map />
</Fragment>
`;

exports[`When API is loading SHOULD show loading 1`] = `
<div>
  Loading
</div>
`;

exports[`When form is submitted AND when the fields are NOT filled SHOULD show error 1`] = `
<Fragment>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <h3>
      Form Demo
    </h3>
    <div
      className="success"
    />
    <FormControl
      errorMessage="Please select one"
      label="Choose userId"
      name="userId"
      onChange={[Function]}
      required={true}
      value=""
    />
    <FormTextInput
      fullWidth={true}
      label="Enter title"
      name="title"
      onChange={[Function]}
      required={true}
      value=""
      variant="outlined"
    />
    <FormTextInput
      fullWidth={true}
      label="Enter body"
      name="body"
      onChange={[Function]}
      required={true}
      value=""
    />
    <ForwardRef(Button)
      disabled={false}
      type="submit"
      variant="outlined"
    >
      Submit
    </ForwardRef(Button)>
  </form>
  <Map />
</Fragment>
`;

exports[`When form is submitted AND when the fields are filled SHOULD show error 1`] = `
<Fragment>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <h3>
      Form Demo
    </h3>
    <div
      className="success"
    />
    <FormControl
      errorMessage="Please select one"
      label="Choose userId"
      name="userId"
      onChange={[Function]}
      required={true}
      value=""
    />
    <FormTextInput
      fullWidth={true}
      label="Enter title"
      name="title"
      onChange={[Function]}
      required={true}
      value=""
      variant="outlined"
    />
    <FormTextInput
      fullWidth={true}
      label="Enter body"
      name="body"
      onChange={[Function]}
      required={true}
      value=""
    />
    <ForwardRef(Button)
      disabled={false}
      type="submit"
      variant="outlined"
    >
      Submit
    </ForwardRef(Button)>
  </form>
  <Map />
</Fragment>
`;
